File: ios/HPTravail/HPTravail/CelestialV2/Core/CelestialCloudDensityV2.swift
import Foundation

// UNRESOLVED is deliberately not inferred to be a low/mid/high altitude.
enum CloudAltitudeV2: Equatable, Sendable { case low, mid, high, unresolved }

struct CloudBandV2: Equatable, Sendable {
    let altitude: CloudAltitudeV2
    let coverage: Double
}

// Qualified weather inputs, not a reconstruction of individual real clouds.
struct CloudAtmosphereStateV2: Equatable, Sendable {
    let totalCoverage: Double
    let lowCoverage: Double?
    let midCoverage: Double?
    let highCoverage: Double?
    let fog: Bool
    let visibilityMeters: Double?

    let bands: [CloudBandV2]

    fileprivate init(
        totalCoverage: Double, lowCoverage: Double?, midCoverage: Double?,
        highCoverage: Double?, fog: Bool, visibilityMeters: Double?
    ) {
        self.totalCoverage = totalCoverage
        self.lowCoverage = lowCoverage
        self.midCoverage = midCoverage
        self.highCoverage = highCoverage
        self.fog = fog
        self.visibilityMeters = visibilityMeters
        let known = [
            highCoverage.map { CloudBandV2(altitude: .high, coverage: $0) },
            midCoverage.map { CloudBandV2(altitude: .mid, coverage: $0) },
            lowCoverage.map { CloudBandV2(altitude: .low, coverage: $0) }
        ].compactMap { $0 }
        self.bands = known.isEmpty ? [CloudBandV2(altitude: .unresolved, coverage: totalCoverage)] : known
    }

    var hasCompleteAltitudeCoverage: Bool {
        lowCoverage != nil && midCoverage != nil && highCoverage != nil
    }
}

enum CelestialCloudAtmosphereV2 {
    // Missing/stale/foreign weather and corrupt percentages are NOT clear sky.
    static func resolve(
        totalCoverage: Double?,
        lowCoverage: Double? = nil,
        midCoverage: Double? = nil,
        highCoverage: Double? = nil,
        fog: Bool = false,
        visibilityMeters: Double? = nil,
        usable: Bool = true
    ) -> CloudAtmosphereStateV2? {
        guard usable, let totalCoverage, validCoverage(totalCoverage) else { return nil }
        for value in [lowCoverage, midCoverage, highCoverage].compactMap({ $0 }) {
            guard validCoverage(value) else { return nil }
        }
        if let visibilityMeters, !visibilityMeters.isFinite || visibilityMeters < 0 { return nil }
        return CloudAtmosphereStateV2(
            totalCoverage: totalCoverage, lowCoverage: lowCoverage,
            midCoverage: midCoverage, highCoverage: highCoverage,
            fog: fog, visibilityMeters: visibilityMeters
        )
    }

    private static func validCoverage(_ value: Double) -> Bool {
        value.isFinite && (0...1).contains(value)
    }
}

// Portable V5 density kernel. Recipes, optical weights and drift are visual
// interpretations, NOT measured microphysics, real cloud positions or wind.
// Callers cache textures and supply a continuous time coordinate. No clock/GPS/API.
// x01 is periodic at the panorama seam; y01 goes from zenith/top to horizon/bottom.
enum CelestialCloudDensityV2 {
    static func sample(
        state: CloudAtmosphereStateV2?,
        x01: Double,
        y01: Double,
        elapsedSeconds: Double,
        octaves: Int = 4
    ) -> Double? {
        guard let state, validCoordinates(x01, y01, elapsedSeconds, octaves) else { return nil }
        var transmission = 1.0
        for band in state.bands {
            guard let alpha = sampleBand(
                band: band, x01: x01, y01: y01,
                elapsedSeconds: elapsedSeconds, octaves: octaves
            ) else { return nil }
            transmission *= 1 - alpha
        }
        if !state.hasCompleteAltitudeCoverage && !state.bands.contains(where: { $0.altitude == .unresolved }) {
            // A partial profile cannot turn aggregate cloudiness into clear sky.
            // Graphical fallback only: no missing altitude is filled or inferred.
            guard let aggregate = sampleBand(
                band: CloudBandV2(altitude: .unresolved, coverage: state.totalCoverage),
                x01: x01, y01: y01, elapsedSeconds: elapsedSeconds, octaves: octaves
            ) else { return nil }
            transmission = min(transmission, 1 - aggregate)
        }
        if state.fog {
            // Fog remains a low continuous veil even with zero total cloud cover.
            let strength = state.visibilityMeters.map { 1 - smooth(100, 2_000, $0) } ?? 0.65
            let veil = (0.15 + 0.70 * strength) * smooth(0.20, 1, y01)
            transmission *= 1 - veil
        }
        return clamp(1 - transmission)
    }

    static func sampleBand(
        band: CloudBandV2,
        x01: Double,
        y01: Double,
        elapsedSeconds: Double,
        octaves: Int = 4
    ) -> Double? {
        guard validCoordinates(x01, y01, elapsedSeconds, octaves),
              band.coverage.isFinite, (0...1).contains(band.coverage) else { return nil }
        if band.coverage == 0 { return 0 }
        let seed: Int
        let stretch: Double
        let opacity: Double
        let drift: Double
        switch band.altitude {
        case .high: seed = 113; stretch = 9; opacity = 0.38; drift = 0.000009
        case .mid: seed = 227; stretch = 2.8; opacity = 0.70; drift = 0.000012
        case .low: seed = 349; stretch = 1.4; opacity = 0.88; drift = 0.000016
        case .unresolved: seed = 491; stretch = 2; opacity = 0.74; drift = 0.000012
        }
        // No short modulo-time reset; translation and deformation stay continuous.
        let x = fract(x01 + elapsedSeconds * drift)
        let y = y01 * stretch + elapsedSeconds * 0.000003
        let warp = noise(x * 4, y * 1.7, 4, seed + 17) - 0.5
        var value = 0.0
        var weight = 0.5
        var totalWeight = 0.0
        var period = 4
        for _ in 0..<octaves {
            value += weight * noise(
                fract(x + warp * 0.08) * Double(period),
                (y + warp * 0.22) * Double(period), period, seed
            )
            totalWeight += weight
            weight *= 0.5
            period *= 2
        }
        let field = clamp((value / totalWeight - 0.5) * 1.65 + 0.5)
        let threshold = 1 - band.coverage
        let softMask = smooth(threshold - 0.18, threshold + 0.18, field)
        let overcastFloor = 0.70 * smooth(0.85, 1, band.coverage)
        return clamp(max(overcastFloor, softMask) * opacity)
    }

    private static func validCoordinates(_ x: Double, _ y: Double, _ seconds: Double, _ octaves: Int) -> Bool {
        x.isFinite && y.isFinite && seconds.isFinite &&
            (-1.0e12...1.0e12).contains(x) && (0...1).contains(y) &&
            (-1.0e12...1.0e12).contains(seconds) && (1...6).contains(octaves)
    }

    private static func noise(_ x: Double, _ y: Double, _ period: Int, _ seed: Int) -> Double {
        let x0 = Int(floor(x))
        let wrappedY = y - floor(y / 1_048_576) * 1_048_576
        let y0 = Int(floor(wrappedY))
        let tx = smooth(0, 1, x - floor(x))
        let ty = smooth(0, 1, wrappedY - floor(wrappedY))
        func h(_ ix: Int, _ iy: Int) -> Double {
            hash((ix % period + period) % period, (iy % 1_048_576 + 1_048_576) % 1_048_576, seed)
        }
        let top = lerp(h(x0, y0), h(x0 + 1, y0), tx)
        let bottom = lerp(h(x0, y0 + 1), h(x0 + 1, y0 + 1), tx)
        return lerp(top, bottom, ty)
    }

    private static func hash(_ x: Int, _ y: Int, _ seed: Int) -> Double {
        // Same wrapping UInt32 hash as the JVM implementation; never Swift Hasher.
        var bits = UInt32(truncatingIfNeeded: x) &* 374761393
        bits = bits &+ (UInt32(truncatingIfNeeded: y) &* 668265263)
        bits = bits &+ (UInt32(truncatingIfNeeded: seed) &* 982451653)
        bits = (bits ^ (bits >> 13)) &* 1274126177
        bits = bits ^ (bits >> 16)
        return Double(bits) / 4294967295
    }

    private static func clamp(_ value: Double) -> Double { min(1, max(0, value)) }
    private static func fract(_ value: Double) -> Double { value - floor(value) }
    private static func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double { a + (b - a) * t }
    private static func smooth(_ a: Double, _ b: Double, _ value: Double) -> Double {
        let t = clamp((value - a) / (b - a))
        return t * t * (3 - 2 * t)
    }
}

File: app/src/main/java/com/amaury/pointage/v2/engine/CelestialCloudDensityV2.kt
package com.amaury.pointage.v2.engine

import kotlin.math.floor
import kotlin.math.max

/** An unresolved altitude is deliberately not inferred from the total cloud cover. */
enum class CloudAltitudeV2 { LOW, MID, HIGH, UNRESOLVED }

data class CloudBandV2(val altitude: CloudAltitudeV2, val coverage: Double)

/** Qualified weather inputs, not a reconstruction of individual real clouds. */
data class CloudAtmosphereStateV2 internal constructor(
    val totalCoverage: Double,
    val lowCoverage: Double?,
    val midCoverage: Double?,
    val highCoverage: Double?,
    val fog: Boolean,
    val visibilityMeters: Double?
) {
    val bands: List<CloudBandV2> = run {
        val known = listOfNotNull(
            highCoverage?.let { CloudBandV2(CloudAltitudeV2.HIGH, it) },
            midCoverage?.let { CloudBandV2(CloudAltitudeV2.MID, it) },
            lowCoverage?.let { CloudBandV2(CloudAltitudeV2.LOW, it) }
        )
        // An aggregate-only response must not become an invented low cloud.
        known.ifEmpty { listOf(CloudBandV2(CloudAltitudeV2.UNRESOLVED, totalCoverage)) }
    }

    val hasCompleteAltitudeCoverage: Boolean
        get() = lowCoverage != null && midCoverage != null && highCoverage != null
}

object CelestialCloudAtmosphereV2 {
    /** Missing/stale/foreign weather and corrupt percentages are NOT clear sky. */
    fun resolve(
        totalCoverage: Double?,
        lowCoverage: Double? = null,
        midCoverage: Double? = null,
        highCoverage: Double? = null,
        fog: Boolean = false,
        visibilityMeters: Double? = null,
        usable: Boolean = true
    ): CloudAtmosphereStateV2? {
        if (!usable || totalCoverage == null || !validCoverage(totalCoverage)) return null
        if (listOf(lowCoverage, midCoverage, highCoverage).any { it != null && !validCoverage(it) }) return null
        if (visibilityMeters != null && (!visibilityMeters.isFinite() || visibilityMeters < 0.0)) return null
        return CloudAtmosphereStateV2(
            totalCoverage, lowCoverage, midCoverage, highCoverage, fog, visibilityMeters
        )
    }

    private fun validCoverage(value: Double): Boolean = value.isFinite() && value in 0.0..1.0
}

/**
 * Portable, deterministic density kernel for the V5 cloud renderer.
 *
 * Recipes, optical weights and drift are visual interpretations, NOT measured
 * microphysics, cloud positions or upper-air wind. No GPS, clock or platform API
 * is read here. Callers cache textures and supply a continuous time coordinate.
 * x01 is periodic at the panorama seam; y01 goes from zenith/top to horizon/bottom.
 */
object CelestialCloudDensityV2 {
    fun sample(
        state: CloudAtmosphereStateV2?,
        x01: Double,
        y01: Double,
        elapsedSeconds: Double,
        octaves: Int = 4
    ): Double? {
        if (state == null || !validCoordinates(x01, y01, elapsedSeconds, octaves)) return null
        var transmission = 1.0
        for (band in state.bands) {
            val alpha = sampleBand(band, x01, y01, elapsedSeconds, octaves) ?: return null
            transmission *= 1.0 - alpha
        }
        if (!state.hasCompleteAltitudeCoverage && state.bands.none { it.altitude == CloudAltitudeV2.UNRESOLVED }) {
            // A partial profile cannot turn aggregate cloudiness into clear sky.
            // This neutral fallback is graphical only: no missing altitude is filled.
            val aggregate = sampleBand(
                CloudBandV2(CloudAltitudeV2.UNRESOLVED, state.totalCoverage),
                x01, y01, elapsedSeconds, octaves
            ) ?: return null
            transmission = minOf(transmission, 1.0 - aggregate)
        }
        if (state.fog) {
            // Fog is a low continuous veil, even if the total cloud cover is zero.
            val strength = state.visibilityMeters?.let { 1.0 - smooth(100.0, 2_000.0, it) } ?: 0.65
            val veil = (0.15 + 0.70 * strength) * smooth(0.20, 1.0, y01)
            transmission *= 1.0 - veil
        }
        return (1.0 - transmission).coerceIn(0.0, 1.0)
    }

    fun sampleBand(
        band: CloudBandV2,
        x01: Double,
        y01: Double,
        elapsedSeconds: Double,
        octaves: Int = 4
    ): Double? {
        if (!validCoordinates(x01, y01, elapsedSeconds, octaves) ||
            !band.coverage.isFinite() || band.coverage !in 0.0..1.0) return null
        if (band.coverage == 0.0) return 0.0

        val seed: Int
        val stretch: Double
        val opacity: Double
        val drift: Double
        when (band.altitude) {
            CloudAltitudeV2.HIGH -> { seed = 113; stretch = 9.0; opacity = 0.38; drift = 0.000009 }
            CloudAltitudeV2.MID -> { seed = 227; stretch = 2.8; opacity = 0.70; drift = 0.000012 }
            CloudAltitudeV2.LOW -> { seed = 349; stretch = 1.4; opacity = 0.88; drift = 0.000016 }
            CloudAltitudeV2.UNRESOLVED -> { seed = 491; stretch = 2.0; opacity = 0.74; drift = 0.000012 }
        }
        // No short modulo-time reset: translation and deformation stay continuous.
        val x = fract(x01 + elapsedSeconds * drift)
        val y = y01 * stretch + elapsedSeconds * 0.000003
        val warp = noise(x * 4.0, y * 1.7, 4, seed + 17) - 0.5
        var value = 0.0
        var weight = 0.5
        var totalWeight = 0.0
        var period = 4
        repeat(octaves) {
            value += weight * noise(
                fract(x + warp * 0.08) * period,
                (y + warp * 0.22) * period,
                period,
                seed
            )
            totalWeight += weight
            weight *= 0.5
            period *= 2
        }
        val field = ((value / totalWeight - 0.5) * 1.65 + 0.5).coerceIn(0.0, 1.0)
        val threshold = 1.0 - band.coverage
        val softMask = smooth(threshold - 0.18, threshold + 0.18, field)
        val overcastFloor = 0.70 * smooth(0.85, 1.0, band.coverage)
        return (max(overcastFloor, softMask) * opacity).coerceIn(0.0, 1.0)
    }

    private fun validCoordinates(x: Double, y: Double, seconds: Double, octaves: Int): Boolean =
        x.isFinite() && y.isFinite() && seconds.isFinite() &&
            x in -1.0e12..1.0e12 && y in 0.0..1.0 &&
            seconds in -1.0e12..1.0e12 && octaves in 1..6

    /** Periodic value noise with smooth interpolation and explicit wrapping hash. */
    private fun noise(x: Double, y: Double, period: Int, seed: Int): Double {
        val x0 = floor(x).toInt()
        // Wrapping a coordinate before Int conversion avoids overflow after long uptime.
        val wrappedY = y - floor(y / 1_048_576.0) * 1_048_576.0
        val y0 = floor(wrappedY).toInt()
        val tx = smooth(0.0, 1.0, x - floor(x))
        val ty = smooth(0.0, 1.0, wrappedY - floor(wrappedY))
        fun h(ix: Int, iy: Int): Double = hash((ix % period + period) % period, (iy % 1_048_576 + 1_048_576) % 1_048_576, seed)
        val top = lerp(h(x0, y0), h(x0 + 1, y0), tx)
        val bottom = lerp(h(x0, y0 + 1), h(x0 + 1, y0 + 1), tx)
        return lerp(top, bottom, ty)
    }

    private fun hash(x: Int, y: Int, seed: Int): Double {
        // JVM Int overflow is intentional; Swift uses the same UInt32 wrapping operations.
        var bits = x * 374761393 + y * 668265263 + seed * 982451653
        bits = (bits xor (bits ushr 13)) * 1274126177
        bits = bits xor (bits ushr 16)
        return (bits.toLong() and 0xffffffffL).toDouble() / 4294967295.0
    }

    private fun fract(value: Double): Double = value - floor(value)
    private fun lerp(a: Double, b: Double, t: Double): Double = a + (b - a) * t
    private fun smooth(a: Double, b: Double, value: Double): Double {
        val t = ((value - a) / (b - a)).coerceIn(0.0, 1.0)
        return t * t * (3.0 - 2.0 * t)
    }
}
